pipeline/update_db.py

```diff
@@ -1,47 +1,85 @@
 import os
+import sys
 from datetime import date
 from argparse import ArgumentParser
 
 import sqlalchemy
 import pandas as pd
 
 
+# expose modules in parent directory
+sys.path.append(os.path.abspath(".."))
+
+from db import Game, Outcome, Session
+
+
 DB_CONN_URI = os.environ.get("DB_CONN_URI", "sqlite:///main.db")
 GAMES_TABLE = os.environ.get("GAMES_TABLE", "games")
 OUTCOMES_TABLE = os.environ.get("OUTCOMES_TABLE", "outcomes")
 DATA_DIR = os.environ.get("UECKER_DATA_DIR", "output")
 
+GAME_COLS = [
+    "id",
+    "home_team",
+    "away_team",
+    "date",
+    "season",
+    "week",
+    "dh",
+    "date_changed",
+]
+OUTCOME_COLS = ["id", "selection_name", "outcome_date", "game_id", "result", "runs"]
+
 
 def load_data(date_str):
     games = pd.read_csv(os.path.join(DATA_DIR, f"schedule_{date_str}.csv"))
+    games.date = pd.to_datetime(games.date).dt.date
     outcomes = pd.read_csv(os.path.join(DATA_DIR, f"outcomes_{date_str}.csv"))
+    outcomes.outcome_date = pd.to_datetime(outcomes.outcome_date).dt.date
     return games, outcomes
 
 
-def get_db_connection():
-    eng = sqlalchemy.create_engine(DB_CONN_URI)
-    return eng
+def update_games(data, session):
+    for _, row in data.iterrows():
+        game = session.query(Game).filter(Game.id == row["id"]).first()
+        # if game not in DB, create new row
+        if not game:
+            game = Game(**row)
+        # if postponed/suspended, flag as date changed and nullify date
+        if row["date_changed"]:
+            game.date_changed = True
+            game.date = None
+        # otherwise ensure date/week/dh are correct
+        else:
+            if game.date != row["date"]:
+                game.date = row["date"]
+                game.week = row["week"]
+            if game.dh != row.dh:
+                game.dh = row["dh"]
+        session.add(game)
 
 
-def update_table(data, table, eng):
-    with eng.begin() as conn:
-        # create temporary table with new data
-        data.to_sql("tmp", conn, index=False, if_exists="replace")
-        # remove those rows from games
-        stmt = f"DELETE FROM {table} WHERE id IN (SELECT id FROM tmp);"
-        conn.execute(stmt)
-        # insert new data into games
-        data.to_sql(table, conn, index=False, if_exists="append")
-        conn.execute("DROP TABLE tmp;")
+def update_outcomes(data, session):
+    for _, row in data.iterrows():
+        oc = session.query(Outcome).filter(Outcome.id == row["id"]).first()
+        # if outcome not in DB, create new row
+        if not oc:
+            oc = Outcome(**row)
+        # if game is complete, update row with results
+        if oc.result != row["result"]:
+            oc.result = row["result"]
+            oc.outcome_date = row["outcome_date"]
+            oc.runs = row["runs"]
+        session.add(oc)
```

Comment:
I'm used to needing a `session.commit()` after adding or modifying ORM objects - is that something you get for free when you use a `with` block?

Reply:
Yep, context managers FTW